src/System.Net.Http/src/System/Net/Http/DiagnosticsHandler.cs

```diff
@@ -0,0 +1,176 @@
+﻿using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    /// <summary>
+    /// DiagnosticHandler notifies DiagnosticSource subscribers about outgoing Http requests
+    /// </summary>
+    internal sealed class DiagnosticsHandler : DelegatingHandler
+    {
+        /// <summary>
+        /// DiagnosticHandler constructor
+        /// </summary>
+        /// <param name="innerHandler">Inner handler: Windows or Unix implementation of HttpMessageHandler. 
+        /// Note that DiagnosticHandler is the latest in the pipeline </param>
+        public DiagnosticsHandler(HttpMessageHandler innerHandler) : base(innerHandler) {}
+
+        protected internal override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+            //do not write to diagnostic source if request is invalid or cancelled,
+            //let inner handler decide what to do with the it
+            if (request == null || cancellationToken.IsCancellationRequested)
+            {
+                return await base.SendAsync(request, cancellationToken);
+            }
+
+            Activity requestActivity = null;
+
+            //Check if DiagnosticSource is enabled for this RequestUri: 
+            //Client code may subscribe to Http Diagnostic Listener with
+            // - no predicate: in this case it will receive ALL events for all outgoing requests which are
+            //      System.Net.Http.Request
+            //      System.Net.Http.Response
+            //      System.Net.Http.Activity.Start
+            //      System.Net.Http.Activity.Stop
+            // - Predicate: Func<string, bool> to filter event names and URIs:
+            //      If user does not want to receive any System.Net.Http.Activity.* events, 
+            //        he may subscribe to Request, Response events only,
+            //      If user wants to filter particular URIs, he may write predicate to achieve it
+            // Also check for Activity.Current: if it's null (subject to change to Activity.IsEnabled) 
+            // this request is not sampled and we don't inject headers
+            if (Activity.Current != null &&
+                s_diagnosticListener.IsEnabled(request.RequestUri.ToString()))
```

Comment:
We have to call IsEnabled for every outgoing request.

I expect user to provide predicate like `uri => instrumentedUris.Any(w => uri.Contains(w))` (or vice versa, provide predicate to decline some Uri). 
The list of Uri-s to enable/disable should be relatively short in average case. Let's say < 10 elements.
The Uri check is also guarded with "sampling" check (`Activity.Current != null`)

If we only call IsEnabled for part of Uri (schema://host:port), it would only allow to change `Contains` or `StartsWith` to `Equal` in predicate. Though, it may disable some scenarios when different services are hidden beside the same gateway and routed with URI.

There might be a story for caching isEnabled result, but I would say it's advanced scenario and I'd prefer user to to it himself  in IsEnabled if it becomes too expensive